{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import gff3\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### This script convert a simple list of gene names to BSU numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Couldnt convert YP_009514011\n",
      "Error: Couldnt convert Novel_transcript\n"
     ]
    }
   ],
   "source": [
    "file = '/Users/andreas/Bacillus/Bioinformatics/Scripts_RNAseq/subtiwiki.gene.export.2024-05-25.tsv'\n",
    "subti = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "\n",
    "file = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/Annotation_files/B_subtilis_GenBank_w_name_and_locus.gff3'\n",
    "genbank = pd.read_csv(file, sep = '\\t', header = 0, comment = '#')\n",
    "\n",
    "### some genes are on the form geneX_geneY\n",
    "### split these genes into two rows; geneX and geneY\n",
    "\n",
    "input_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits'\n",
    "output_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits/BSU'\n",
    "\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    df = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "    gene_base = []\n",
    "    for gene in df['Gene'].tolist():\n",
    "        gene_base.append(gene.split('.')[0])\n",
    "    df['Gene_base'] = gene_base\n",
    "    df = df.drop_duplicates(['Gene_base'])\n",
    "    df.to_csv(file, sep = '\\t', header = True, index = False)\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    all = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    temp_gene_names = []\n",
    "    for name in all['Gene_base'].tolist():\n",
    "        temp_gene_names.append(name.replace('BSU_', 'BSU^').replace('NP_', 'NP^').replace('YP_', 'YP^').replace('Novel_transcript', 'Novel^transcript'))\n",
    "    all['Gene_base'] = temp_gene_names\n",
    "\n",
    "    _df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each row in the original DataFrame\n",
    "    for index, row in all.iterrows():\n",
    "        # Check if 'Gene_base' contains an underscore\n",
    "        if '_' in row['Gene_base']:\n",
    "            # Split the 'Gene_base' value into a list of genes\n",
    "            gene_list = row['Gene_base'].split('_')\n",
    "            # Create new rows for each gene in gene_list\n",
    "            for gene in gene_list:\n",
    "                new_row = pd.DataFrame({\n",
    "                    'Gene': [gene],\n",
    "                    'Gene_base': [gene]\n",
    "                })\n",
    "                _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "        ### replace '@' with ' ' in gene names\n",
    "        elif '@' in row['Gene_base']:\n",
    "            new_row = pd.DataFrame({\n",
    "                    'Gene_base': [row['Gene_base'].replace('@', ' ')]\n",
    "                })\n",
    "            _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "\n",
    "        else:\n",
    "            # If nothing else, keep the row as is\n",
    "            _df = pd.concat([_df, all[all.index == index]], ignore_index=True)\n",
    "    \n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    revert_gene_names = []\n",
    "    for name in _df['Gene_base'].tolist():\n",
    "        revert_gene_names.append(name.replace('BSU^', 'BSU_').replace('NP^', 'NP_').replace('YP^', 'YP_').replace('Novel^transcript', 'Novel_transcript'))\n",
    "    _df['Gene_base'] = revert_gene_names\n",
    "\n",
    "\n",
    "    ### convert gene names to BSU_numbers using SubtiWiki and Genbank\n",
    "    bsu_numbers = []\n",
    "    for gene in _df['Gene_base'].tolist():\n",
    "        try:\n",
    "            bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "            continue\n",
    "        except IndexError:\n",
    "            x = 0\n",
    "            for index, row in subti.iterrows():\n",
    "                try:\n",
    "                    if gene in row['synonyms']:\n",
    "                        bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "                        x += 1\n",
    "                        break\n",
    "                ## for na values\n",
    "                except TypeError:\n",
    "                    pass\n",
    "            if x == 0:\n",
    "                try:\n",
    "                    bsu_numbers.append(genbank['locusTag'][genbank['name'] == gene].iloc[0])\n",
    "                    continue\n",
    "                except IndexError:\n",
    "                    print(f'Error: Couldnt convert {gene}')\n",
    "                    bsu_numbers.append('no_locusTag')\n",
    "    \n",
    "    _df['locusTag'] = bsu_numbers\n",
    "    _df.drop_duplicates(['locusTag'])\n",
    "    _df.to_csv(os.path.join(output_folder, os.path.basename(file)), sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ypoP</td>\n",
       "      <td>ypoP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpoC</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>resA</td>\n",
       "      <td>resA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ybgB</td>\n",
       "      <td>ybgB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gltT</td>\n",
       "      <td>gltT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>folEB</td>\n",
       "      <td>folEB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ypuD</td>\n",
       "      <td>ypuD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ftsE</td>\n",
       "      <td>ftsE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rpsR</td>\n",
       "      <td>rpsR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nfeDB</td>\n",
       "      <td>nfeDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rpsP</td>\n",
       "      <td>rpsP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fliF</td>\n",
       "      <td>fliF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ypgR</td>\n",
       "      <td>ypgR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cysJ</td>\n",
       "      <td>cysJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>metQ</td>\n",
       "      <td>metQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cydA</td>\n",
       "      <td>cydA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S19</td>\n",
       "      <td>S19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ybeC</td>\n",
       "      <td>ybeC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hemL</td>\n",
       "      <td>hemL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ppiB</td>\n",
       "      <td>ppiB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rny</td>\n",
       "      <td>rny</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>mbl</td>\n",
       "      <td>mbl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rosA</td>\n",
       "      <td>rosA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>S346</td>\n",
       "      <td>S346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>S1072</td>\n",
       "      <td>S1072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>znuA</td>\n",
       "      <td>znuA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wapA</td>\n",
       "      <td>wapA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>mswB</td>\n",
       "      <td>mswB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>S471</td>\n",
       "      <td>S471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>metE</td>\n",
       "      <td>metE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>sufC</td>\n",
       "      <td>sufC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>yuiB</td>\n",
       "      <td>yuiB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>mswD</td>\n",
       "      <td>mswD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene Gene_base\n",
       "0    ypoP      ypoP\n",
       "1    rpoC      rpoC\n",
       "2    resA      resA\n",
       "3    ybgB      ybgB\n",
       "4    gltT      gltT\n",
       "5   folEB     folEB\n",
       "6    ypuD      ypuD\n",
       "7    ftsE      ftsE\n",
       "8    rpsR      rpsR\n",
       "9   nfeDB     nfeDB\n",
       "10   rpsP      rpsP\n",
       "11   fliF      fliF\n",
       "12   ypgR      ypgR\n",
       "13   cysJ      cysJ\n",
       "14   metQ      metQ\n",
       "15   cydA      cydA\n",
       "16    S19       S19\n",
       "17   ybeC      ybeC\n",
       "18   hemL      hemL\n",
       "19   ppiB      ppiB\n",
       "20    rny       rny\n",
       "21    mbl       mbl\n",
       "22   rosA      rosA\n",
       "23   S346      S346\n",
       "24  S1072     S1072\n",
       "25   znuA      znuA\n",
       "26   wapA      wapA\n",
       "27   mswB      mswB\n",
       "28   S471      S471\n",
       "29   metE      metE\n",
       "30   sufC      sufC\n",
       "31   yuiB      yuiB\n",
       "32   mswD      mswD"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>locusTag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>4215606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=NC_000964.3:1..4215606;Dbxref=taxon:224308;...</td>\n",
       "      <td>no_name</td>\n",
       "      <td>no_locusTag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>410</td>\n",
       "      <td>1750</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-BSU_00010;Dbxref=GeneID:939978;Name=dn...</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>BSU_00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>410</td>\n",
       "      <td>1750</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_387882.1;Parent=gene-BSU_00010;Dbxre...</td>\n",
       "      <td>NP_387882.1</td>\n",
       "      <td>BSU_00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>1939</td>\n",
       "      <td>3075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-BSU_00020;Dbxref=GeneID:939970;Name=dn...</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>BSU_00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1939</td>\n",
       "      <td>3075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_387883.1;Parent=gene-BSU_00020;Dbxre...</td>\n",
       "      <td>NP_387883.1</td>\n",
       "      <td>BSU_00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9283</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4213823</td>\n",
       "      <td>4214608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_391984.1;Parent=gene-BSU_41040;Dbxre...</td>\n",
       "      <td>NP_391984.1</td>\n",
       "      <td>BSU_41040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9284</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>4214753</td>\n",
       "      <td>4215103</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-BSU_41050;Dbxref=GeneID:937930;Name=rn...</td>\n",
       "      <td>rnpA</td>\n",
       "      <td>BSU_41050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9285</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4214753</td>\n",
       "      <td>4215103</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_391985.1;Parent=gene-BSU_41050;Dbxre...</td>\n",
       "      <td>NP_391985.1</td>\n",
       "      <td>BSU_41050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9286</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>4215255</td>\n",
       "      <td>4215389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-BSU_41060;Dbxref=GeneID:937942;Name=rp...</td>\n",
       "      <td>rpmH</td>\n",
       "      <td>BSU_41060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9287</th>\n",
       "      <td>NC_000964.3</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4215255</td>\n",
       "      <td>4215389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_391986.1;Parent=gene-BSU_41060;Dbxre...</td>\n",
       "      <td>NP_391986.1</td>\n",
       "      <td>BSU_41060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9288 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqID  source feature    start      end score strand phase  \\\n",
       "0     NC_000964.3  RefSeq  region        1  4215606     .      +     .   \n",
       "1     NC_000964.3  RefSeq    gene      410     1750     .      +     .   \n",
       "2     NC_000964.3  RefSeq     CDS      410     1750     .      +     0   \n",
       "3     NC_000964.3  RefSeq    gene     1939     3075     .      +     .   \n",
       "4     NC_000964.3  RefSeq     CDS     1939     3075     .      +     0   \n",
       "...           ...     ...     ...      ...      ...   ...    ...   ...   \n",
       "9283  NC_000964.3  RefSeq     CDS  4213823  4214608     .      -     0   \n",
       "9284  NC_000964.3  RefSeq    gene  4214753  4215103     .      -     .   \n",
       "9285  NC_000964.3  RefSeq     CDS  4214753  4215103     .      -     0   \n",
       "9286  NC_000964.3  RefSeq    gene  4215255  4215389     .      -     .   \n",
       "9287  NC_000964.3  RefSeq     CDS  4215255  4215389     .      -     0   \n",
       "\n",
       "                                             attributes         name  \\\n",
       "0     ID=NC_000964.3:1..4215606;Dbxref=taxon:224308;...      no_name   \n",
       "1     ID=gene-BSU_00010;Dbxref=GeneID:939978;Name=dn...         dnaA   \n",
       "2     ID=cds-NP_387882.1;Parent=gene-BSU_00010;Dbxre...  NP_387882.1   \n",
       "3     ID=gene-BSU_00020;Dbxref=GeneID:939970;Name=dn...         dnaN   \n",
       "4     ID=cds-NP_387883.1;Parent=gene-BSU_00020;Dbxre...  NP_387883.1   \n",
       "...                                                 ...          ...   \n",
       "9283  ID=cds-NP_391984.1;Parent=gene-BSU_41040;Dbxre...  NP_391984.1   \n",
       "9284  ID=gene-BSU_41050;Dbxref=GeneID:937930;Name=rn...         rnpA   \n",
       "9285  ID=cds-NP_391985.1;Parent=gene-BSU_41050;Dbxre...  NP_391985.1   \n",
       "9286  ID=gene-BSU_41060;Dbxref=GeneID:937942;Name=rp...         rpmH   \n",
       "9287  ID=cds-NP_391986.1;Parent=gene-BSU_41060;Dbxre...  NP_391986.1   \n",
       "\n",
       "         locusTag  \n",
       "0     no_locusTag  \n",
       "1       BSU_00010  \n",
       "2       BSU_00010  \n",
       "3       BSU_00020  \n",
       "4       BSU_00020  \n",
       "...           ...  \n",
       "9283    BSU_41040  \n",
       "9284    BSU_41050  \n",
       "9285    BSU_41050  \n",
       "9286    BSU_41060  \n",
       "9287    BSU_41060  \n",
       "\n",
       "[9288 rows x 11 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RNAseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
