{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import gff3\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_390867\n"
     ]
    }
   ],
   "source": [
    "file = '/Users/andreas/Bacillus/Bioinformatics/Scripts_RNAseq/subtiwiki.gene.export.2024-05-25.tsv'\n",
    "subti = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "\n",
    "# file = '/Users/andreas/Bacillus/Bioinformatics/Scripts_RNAseq/GenBank.gff3'\n",
    "# genbank = pd.read_csv(file, sep = '\\t', names = gff3.header, comment = '#')\n",
    "file = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/Annotation_files/B_subtilis_GenBank_w_name_and_locus.gff3'\n",
    "genbank = pd.read_csv(file, sep = '\\t', header = 0, comment = '#')\n",
    "\n",
    "### some genes are on the form geneX_geneY\n",
    "### split these genes into two rows; geneX and geneY\n",
    "\n",
    "input_folder = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/DEG'\n",
    "output_folder = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/DEG_significant'\n",
    "\n",
    "# input_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits'\n",
    "# ouput_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits/BSU'\n",
    "\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    df = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "    gene_base = []\n",
    "    for gene in df['Gene'].tolist():\n",
    "        if gene.find('NP_') == 0 and gene.count('_') == 1:\n",
    "            gene_base.append(gene.split('.')[0] + '.' + gene.split('.')[1])\n",
    "        else:\n",
    "            gene_base.append(gene.split('.')[0])\n",
    "    df['Gene_base'] = gene_base\n",
    "    df = df.sort_values(['padj'], ascending = True)\n",
    "    df = df.drop_duplicates(['Gene_base'])\n",
    "    df.to_csv(file, sep = '\\t', header = True, index = False)\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    all = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    temp_gene_names = []\n",
    "    for name in all['Gene_base'].tolist():\n",
    "        temp_gene_names.append(name.replace('BSU_', 'BSU^').replace('NP_', 'NP^').replace('YP_', 'YP^').replace('Novel_transcript', 'Novel^transcript'))\n",
    "    all['Gene_base'] = temp_gene_names\n",
    "\n",
    "    _df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each row in the original DataFrame\n",
    "    for index, row in all.iterrows():\n",
    "        # Check if 'Gene_base' contains an underscore\n",
    "        if '_' in row['Gene_base']:\n",
    "            # Split the 'Gene_base' value into a list of genes\n",
    "            gene_list = row['Gene_base'].split('_')\n",
    "            # Create new rows for each gene in gene_list\n",
    "            for gene in gene_list:\n",
    "                new_row = pd.DataFrame({\n",
    "                    'Gene_base': [gene],\n",
    "                    'baseMean': [row['baseMean']],\n",
    "                    'log2FoldChange': [row['log2FoldChange']],\n",
    "                    'lfcSE': [row['lfcSE']],\n",
    "                    'stat': [row['stat']],\n",
    "                    'pvalue': [row['pvalue']],\n",
    "                    'padj': [row['padj']]\n",
    "                })\n",
    "                _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "        ### replace '@' with ' ' in gene names\n",
    "        elif '@' in row['Gene_base']:\n",
    "            new_row = pd.DataFrame({\n",
    "                    'Gene_base': [row['Gene_base'].replace('@', ' ')],\n",
    "                    'baseMean': [row['baseMean']],\n",
    "                    'log2FoldChange': [row['log2FoldChange']],\n",
    "                    'lfcSE': [row['lfcSE']],\n",
    "                    'stat': [row['stat']],\n",
    "                    'pvalue': [row['pvalue']],\n",
    "                    'padj': [row['padj']]\n",
    "                })\n",
    "            _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "\n",
    "        else:\n",
    "            # If nothing else, keep the row as is\n",
    "            _df = pd.concat([_df, all[all.index == index]], ignore_index=True)\n",
    "    \n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    revert_gene_names = []\n",
    "    for name in _df['Gene_base'].tolist():\n",
    "        revert_gene_names.append(name.replace('BSU^', 'BSU_').replace('NP^', 'NP_').replace('YP^', 'YP_').replace('Novel^transcript', 'Novel_transcript'))\n",
    "    _df['Gene_base'] = revert_gene_names\n",
    "\n",
    "    ### make list of DEG\n",
    "    sig_genes = list(set(_df['Gene_base'][_df['padj'] <= 0.05].tolist()))\n",
    "\n",
    "    ### convert gene names to BSU_numbers\n",
    "    bsu_numbers = []\n",
    "    # for gene in sig_genes:\n",
    "    #     try:\n",
    "    #         bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "    #     except IndexError:\n",
    "    #         x = 0\n",
    "    #         for index, row in subti.iterrows():\n",
    "    #             try:\n",
    "    #                 if gene in row['synonyms']:\n",
    "    #                     bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "    #                     x += 1\n",
    "    #                     break\n",
    "    #             ## for na values\n",
    "    #             except TypeError:\n",
    "    #                 pass\n",
    "    #         if x == 0:\n",
    "    #             print(f'Error: Couldnt convert {gene}')\n",
    "    for gene in sig_genes:\n",
    "        try:\n",
    "            bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "            continue\n",
    "        except IndexError:\n",
    "            x = 0\n",
    "            for index, row in subti.iterrows():\n",
    "                try:\n",
    "                    if gene in row['synonyms']:\n",
    "                        bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "                        x += 1\n",
    "                        break\n",
    "                ## for na values\n",
    "                except TypeError:\n",
    "                    pass\n",
    "            if x == 0:\n",
    "                try:\n",
    "                    bsu_numbers.append(genbank['locusTag'][genbank['name'] == gene].iloc[0])\n",
    "                    continue\n",
    "                except IndexError:\n",
    "                    print(f'Error: Couldnt convert {gene}')\n",
    "                    bsu_numbers.append('no_locusTag')\n",
    "    \n",
    "    bsunumbers = [i.replace('_', '') for i in bsu_numbers]\n",
    "\n",
    "    ### convert bsu_numbers to genbank names\n",
    "    genbank_names = []\n",
    "    for bsu in bsu_numbers:\n",
    "        for index, att in enumerate(genbank['attributes'].tolist()):\n",
    "            x = 0\n",
    "            if bsu in att:\n",
    "                att_list = att.split(';')\n",
    "                for a in att_list:\n",
    "                    if a.find('gene=') == 0:\n",
    "                        genbank_names.append(a[5:])\n",
    "                        x += 1\n",
    "                        break\n",
    "            if x == 1:\n",
    "                break\n",
    "    \n",
    "    ### Write list to .txt file\n",
    "    # Specify the file name\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in list(set(genbank_names)):\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n",
    "\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes_BSU_Numbers.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in list(set(bsu_numbers)):\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n",
    "\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes_BSUNumbers_no_underscore.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in list(set(bsunumbers)):\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Couldnt convert tuaAn\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_390867\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_388994\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert rna-BSU_misc\n",
      "Error: Couldnt convert YP_054580\n",
      "Error: Couldnt convert NP_389652\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert rna-BSU_misc\n",
      "Error: Couldnt convert NP_388994\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert NP_391764\n",
      "Error: Couldnt convert rna-BSU_misc\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_389505\n",
      "Error: Couldnt convert NP_389652\n",
      "Error: Couldnt convert NP_389573\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert NP_389573\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert NP_390648\n",
      "Error: Couldnt convert NP_391764\n",
      "Error: Couldnt convert YP_054580\n",
      "Error: Couldnt convert NP_391764\n",
      "Error: Couldnt convert NP_390648\n",
      "Error: Couldnt convert NP_390648\n",
      "Error: Couldnt convert NP_390648\n",
      "Error: Couldnt convert NP_390867\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert mgtE@riboswitch\n",
      "Error: Couldnt convert yusYc\n",
      "Error: Couldnt convert yusYn\n",
      "Error: Couldnt convert NP_391409\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_389886\n",
      "Error: Couldnt convert NP_390504\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_388994\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_389888\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert id-BSU_20060\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert yqbNn\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert spsMn\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_389886\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert ydhUn\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert NP_389888\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n"
     ]
    }
   ],
   "source": [
    "### convert gene names to BSU_numbers on whole of DEG file\n",
    "DEG_folder = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/DEG'\n",
    "\n",
    "for file in glob.glob(DEG_folder + '/*.txt'):\n",
    "    all = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    temp_gene_names = []\n",
    "    for name in all['Gene_base'].tolist():\n",
    "        temp_gene_names.append(name.replace('BSU_', 'BSU^').replace('NP_', 'NP^').replace('YP_', 'YP^').replace('Novel_transcript', 'Novel^transcript'))\n",
    "    all['Gene_base'] = temp_gene_names\n",
    "\n",
    "    _df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each row in the original DataFrame\n",
    "    for index, row in all.iterrows():\n",
    "        # Check if 'Gene_base' contains an underscore\n",
    "        if '_' in row['Gene_base']:\n",
    "            # Split the 'Gene_base' value into a list of genes\n",
    "            gene_list = row['Gene_base'].split('_')\n",
    "            # Create new rows for each gene in gene_list\n",
    "            for gene in gene_list:\n",
    "                new_row = pd.DataFrame({\n",
    "                    'Gene': [row['Gene']],\n",
    "                    'Gene_base': [gene],\n",
    "                    'baseMean': [row['baseMean']],\n",
    "                    'log2FoldChange': [row['log2FoldChange']],\n",
    "                    'lfcSE': [row['lfcSE']],\n",
    "                    'stat': [row['stat']],\n",
    "                    'pvalue': [row['pvalue']],\n",
    "                    'padj': [row['padj']]\n",
    "                })\n",
    "                _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "        ### replace '@' with ' ' in gene names\n",
    "        elif '@' in row['Gene_base']:\n",
    "            new_row = pd.DataFrame({\n",
    "                    'Gene': [row['Gene']],\n",
    "                    'Gene_base': [row['Gene_base'].replace('@', ' ')],\n",
    "                    'baseMean': [row['baseMean']],\n",
    "                    'log2FoldChange': [row['log2FoldChange']],\n",
    "                    'lfcSE': [row['lfcSE']],\n",
    "                    'stat': [row['stat']],\n",
    "                    'pvalue': [row['pvalue']],\n",
    "                    'padj': [row['padj']]\n",
    "                })\n",
    "            _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "\n",
    "        else:\n",
    "            # If nothing else, keep the row as is\n",
    "            _df = pd.concat([_df, all[all.index == index]], ignore_index=True)\n",
    "    \n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    revert_gene_names = []\n",
    "    for name in _df['Gene_base'].tolist():\n",
    "        revert_gene_names.append(name.replace('BSU^', 'BSU_').replace('NP^', 'NP_').replace('YP^', 'YP_').replace('Novel^transcript', 'Novel_transcript'))\n",
    "    _df['Gene_base'] = revert_gene_names\n",
    "\n",
    "\n",
    "    bsu_numbers = []\n",
    "    for gene in _df['Gene_base'].tolist():\n",
    "        try:\n",
    "            bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "        except IndexError:\n",
    "            x = 0\n",
    "            for index, row in subti.iterrows():\n",
    "                try:\n",
    "                    if gene in row['synonyms']:\n",
    "                        bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "                        x += 1\n",
    "                        break\n",
    "                ## for na values\n",
    "                except TypeError:\n",
    "                    pass\n",
    "            if x == 0:\n",
    "                print(f'Error: Couldnt convert {gene}')\n",
    "                bsu_numbers.append(gene)\n",
    "    \n",
    "    _df['locus'] = bsu_numbers\n",
    "    _df.to_csv(file, sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>sig</th>\n",
       "      <th>Gene_base</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ygxB.2</td>\n",
       "      <td>3247.827950</td>\n",
       "      <td>-1.096988</td>\n",
       "      <td>0.133340</td>\n",
       "      <td>-8.226980</td>\n",
       "      <td>1.919932e-16</td>\n",
       "      <td>4.980303e-13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ygxB</td>\n",
       "      <td>BSU_09390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ydaG.2</td>\n",
       "      <td>7111.139000</td>\n",
       "      <td>-0.717854</td>\n",
       "      <td>0.086768</td>\n",
       "      <td>-8.273288</td>\n",
       "      <td>1.303140e-16</td>\n",
       "      <td>4.980303e-13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ydaG</td>\n",
       "      <td>BSU_04220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ydbD.2</td>\n",
       "      <td>1382.085553</td>\n",
       "      <td>-0.907377</td>\n",
       "      <td>0.113316</td>\n",
       "      <td>-8.007472</td>\n",
       "      <td>1.170903e-15</td>\n",
       "      <td>2.277991e-12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ydbD</td>\n",
       "      <td>BSU_04430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yflT.2</td>\n",
       "      <td>5008.471201</td>\n",
       "      <td>-0.801242</td>\n",
       "      <td>0.100872</td>\n",
       "      <td>-7.943139</td>\n",
       "      <td>1.971271e-15</td>\n",
       "      <td>3.068086e-12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yflT</td>\n",
       "      <td>BSU_07550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ohrB.2</td>\n",
       "      <td>3527.509671</td>\n",
       "      <td>-0.721755</td>\n",
       "      <td>0.092329</td>\n",
       "      <td>-7.817180</td>\n",
       "      <td>5.401965e-15</td>\n",
       "      <td>7.006349e-12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ohrB</td>\n",
       "      <td>BSU_13160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10688</th>\n",
       "      <td>manP_manA.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>manA</td>\n",
       "      <td>BSU_12020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10689</th>\n",
       "      <td>S666.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S666</td>\n",
       "      <td>new_1925540_1925623_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10690</th>\n",
       "      <td>S678.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S678</td>\n",
       "      <td>new_1949513_1949584_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10691</th>\n",
       "      <td>BSU_19915.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSU_19915</td>\n",
       "      <td>BSU_19915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10692</th>\n",
       "      <td>yuzL.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yuzL</td>\n",
       "      <td>BSU_32849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10693 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene     baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0           ygxB.2  3247.827950       -1.096988  0.133340 -8.226980   \n",
       "1           ydaG.2  7111.139000       -0.717854  0.086768 -8.273288   \n",
       "2           ydbD.2  1382.085553       -0.907377  0.113316 -8.007472   \n",
       "3           yflT.2  5008.471201       -0.801242  0.100872 -7.943139   \n",
       "4           ohrB.2  3527.509671       -0.721755  0.092329 -7.817180   \n",
       "...            ...          ...             ...       ...       ...   \n",
       "10688  manP_manA.1     0.000000             NaN       NaN       NaN   \n",
       "10689       S666.1     0.000000             NaN       NaN       NaN   \n",
       "10690       S678.1     0.000000             NaN       NaN       NaN   \n",
       "10691  BSU_19915.1     0.000000             NaN       NaN       NaN   \n",
       "10692       yuzL.1     0.000000             NaN       NaN       NaN   \n",
       "\n",
       "             pvalue          padj  sig  Gene_base                  locus  \n",
       "0      1.919932e-16  4.980303e-13  1.0       ygxB              BSU_09390  \n",
       "1      1.303140e-16  4.980303e-13  1.0       ydaG              BSU_04220  \n",
       "2      1.170903e-15  2.277991e-12  1.0       ydbD              BSU_04430  \n",
       "3      1.971271e-15  3.068086e-12  1.0       yflT              BSU_07550  \n",
       "4      5.401965e-15  7.006349e-12  1.0       ohrB              BSU_13160  \n",
       "...             ...           ...  ...        ...                    ...  \n",
       "10688           NaN           NaN  NaN       manA              BSU_12020  \n",
       "10689           NaN           NaN  NaN       S666  new_1925540_1925623_c  \n",
       "10690           NaN           NaN  NaN       S678  new_1949513_1949584_c  \n",
       "10691           NaN           NaN  NaN  BSU_19915              BSU_19915  \n",
       "10692           NaN           NaN  NaN       yuzL              BSU_32849  \n",
       "\n",
       "[10693 rows x 10 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Couldnt convert YP_009514003\n",
      "Error: Couldnt convert YP_009513941\n",
      "Error: Couldnt convert YP_009514007\n",
      "Error: Couldnt convert YP_009513997\n",
      "Error: Couldnt convert YP_009513943\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert YP_009514013\n",
      "Error: Couldnt convert YP_009513997\n",
      "Error: Couldnt convert YP_009514003\n",
      "Error: Couldnt convert YP_009514007\n",
      "Error: Couldnt convert Novel_transcript\n",
      "Error: Couldnt convert YP_009513989\n"
     ]
    }
   ],
   "source": [
    "### prepare CRAC to GO analysis\n",
    "file = '/Users/andreas/Bacillus/Bioinformatics/Scripts_RNAseq/subtiwiki.gene.export.2024-05-25.tsv'\n",
    "subti = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "\n",
    "# file = '/Users/andreas/Bacillus/Bioinformatics/Scripts_RNAseq/GenBank.gff3'\n",
    "# genbank = pd.read_csv(file, sep = '\\t', names = gff3.header, comment = '#')\n",
    "file = '/Users/andreas/Bacillus/Bioinformatics/RNAseq_jag_khpa_kre/X204SC22071444-Z01-F001/Analysis/Annotation_files/B_subtilis_GenBank_w_name_and_locus.gff3'\n",
    "genbank = pd.read_csv(file, sep = '\\t', header = 0, comment = '#')\n",
    "\n",
    "### some genes are on the form geneX_geneY\n",
    "### split these genes into two rows; geneX and geneY\n",
    "\n",
    "input_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/ratios'\n",
    "output_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/ratios/prepared'\n",
    "\n",
    "# input_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits'\n",
    "# ouput_folder = '/Users/andreas/Bacillus/Bioinformatics/CRAC_jag_khpA_kre_30thMay2024/DPW/z_scores/Top_CRAC_hits/BSU'\n",
    "\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    df = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "    gene_base = []\n",
    "    for gene in df['Gene'].tolist():\n",
    "        if gene.find('NP_') == 0 and gene.count('_') == 1:\n",
    "            gene_base.append(gene.split('.')[0] + '.' + gene.split('.')[1])\n",
    "        else:\n",
    "            gene_base.append(gene.split('.')[0])\n",
    "    df['Gene_base'] = gene_base\n",
    "    df = df.drop_duplicates(['Gene_base'])\n",
    "    df.to_csv(file, sep = '\\t', header = True, index = False)\n",
    "\n",
    "for file in glob.glob(input_folder + '/*.txt'):\n",
    "    all = pd.read_csv(file, sep = '\\t', header = 0)\n",
    "\n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    temp_gene_names = []\n",
    "    for name in all['Gene_base'].tolist():\n",
    "        temp_gene_names.append(name.replace('BSU_', 'BSU^').replace('NP_', 'NP^').replace('YP_', 'YP^').replace('Novel_transcript', 'Novel^transcript'))\n",
    "    all['Gene_base'] = temp_gene_names\n",
    "\n",
    "    _df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each row in the original DataFrame\n",
    "    for index, row in all.iterrows():\n",
    "        # Check if 'Gene_base' contains an underscore\n",
    "        if '_' in row['Gene_base']:\n",
    "            # Split the 'Gene_base' value into a list of genes\n",
    "            gene_list = row['Gene_base'].split('_')\n",
    "            # Create new rows for each gene in gene_list\n",
    "            for gene in gene_list:\n",
    "                new_row = pd.DataFrame({\n",
    "                    'Gene_base': [gene]\n",
    "                })\n",
    "                _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "        ### replace '@' with ' ' in gene names\n",
    "        elif '@' in row['Gene_base']:\n",
    "            new_row = pd.DataFrame({\n",
    "                    'Gene_base': [row['Gene_base'].replace('@', ' ')]\n",
    "                })\n",
    "            _df = pd.concat([_df, new_row], ignore_index=True)\n",
    "\n",
    "        else:\n",
    "            # If nothing else, keep the row as is\n",
    "            all_temp = pd.DataFrame()\n",
    "            all_temp['Gene_base'] = all['Gene_base'][all.index == index].tolist()\n",
    "            _df = pd.concat([_df, all_temp], ignore_index=True)\n",
    "    \n",
    "    \n",
    "    ### replace 'BSU_' with 'BSU^' \n",
    "    ### This way we can seperate transcripts by _ followingly\n",
    "    revert_gene_names = []\n",
    "    for name in _df['Gene_base'].tolist():\n",
    "        revert_gene_names.append(name.replace('BSU^', 'BSU_').replace('NP^', 'NP_').replace('YP^', 'YP_').replace('Novel^transcript', 'Novel_transcript'))\n",
    "    _df['Gene_base'] = revert_gene_names\n",
    "    _df = _df.drop_duplicates(['Gene_base'])\n",
    "\n",
    "    ### make list of DEG\n",
    "    sig_genes = _df['Gene_base'].tolist()\n",
    "\n",
    "    ### convert gene names to BSU_numbers\n",
    "    bsu_numbers = []\n",
    "    # for gene in sig_genes:\n",
    "    #     try:\n",
    "    #         bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "    #     except IndexError:\n",
    "    #         x = 0\n",
    "    #         for index, row in subti.iterrows():\n",
    "    #             try:\n",
    "    #                 if gene in row['synonyms']:\n",
    "    #                     bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "    #                     x += 1\n",
    "    #                     break\n",
    "    #             ## for na values\n",
    "    #             except TypeError:\n",
    "    #                 pass\n",
    "    #         if x == 0:\n",
    "    #             print(f'Error: Couldnt convert {gene}')\n",
    "    for gene in sig_genes:\n",
    "        try:\n",
    "            bsu_numbers.append(subti['locus'][subti['title'] == gene].iloc[0])\n",
    "            continue\n",
    "        except IndexError:\n",
    "            x = 0\n",
    "            for index, row in subti.iterrows():\n",
    "                try:\n",
    "                    if gene in row['synonyms']:\n",
    "                        bsu_numbers.append(subti['locus'][subti.index == index].iloc[0])\n",
    "                        x += 1\n",
    "                        break\n",
    "                ## for na values\n",
    "                except TypeError:\n",
    "                    pass\n",
    "            if x == 0:\n",
    "                try:\n",
    "                    bsu_numbers.append(genbank['locusTag'][genbank['name'] == gene].iloc[0])\n",
    "                    continue\n",
    "                except IndexError:\n",
    "                    print(f'Error: Couldnt convert {gene}')\n",
    "                    bsu_numbers.append('no_locusTag')\n",
    "    \n",
    "    bsunumbers = [i.replace('_', '') for i in bsu_numbers]\n",
    "\n",
    "    ### convert bsu_numbers to genbank names\n",
    "    genbank_names = []\n",
    "    for bsu in bsu_numbers:\n",
    "        for index, att in enumerate(genbank['attributes'].tolist()):\n",
    "            x = 0\n",
    "            if bsu in att:\n",
    "                att_list = att.split(';')\n",
    "                for a in att_list:\n",
    "                    if a.find('gene=') == 0:\n",
    "                        genbank_names.append(a[5:])\n",
    "                        x += 1\n",
    "                        break\n",
    "            if x == 1:\n",
    "                break\n",
    "    \n",
    "    ### Write list to .txt file\n",
    "    # Specify the file name\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in genbank_names:\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n",
    "\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes_BSU_Numbers.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in bsu_numbers:\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n",
    "\n",
    "    file_name = f\"{output_folder}/{os.path.basename(file)}_sig_genes_BSUNumbers_no_underscore.txt\"\n",
    "\n",
    "    # Open the file in write mode\n",
    "    with open(file_name, \"w\") as filex:\n",
    "        # Iterate through each item in the list\n",
    "        for item in bsunumbers:\n",
    "            # Write the item followed by a newline character\n",
    "            filex.write(f\"{item}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
